File: scripts/check_mcp_config.js
#!/usr/bin/env node
/**
 * Lightweight static check for .cursor/mcp.json (no secrets printed).
 */

const fs = require("fs");
const path = require("path");

const ROOT = path.resolve(__dirname, "..");
const DEFAULT_MCP_PATH = path.join(ROOT, ".cursor", "mcp.json");

const SUPPORTED_SERVERS = [
  "chrome-devtools",
  "context7",
  "filesystem",
  "github",
  "playwright",
];

const DANGEROUS_PATH_PATTERNS = [
  /^\/$/,
  /^[A-Za-z]:\\?$/,
  /^~\/ ?$/,
  /^\/Users\/[^/]+\/?$/,
  /^\/home\/[^/]+\/?$/,
  /^C:\\Users\\[^\\]+\/?$/i,
  /^\/Volumes\/?$/,
];

const SENSITIVE_ENV_KEY = /(token|secret|password|api[_-]?key|cookie|session)/i;

function isDangerousPath(arg) {
  const normalized = String(arg).trim().replace(/\\/g, "/");
  if (["/", "//", "C:", "C:/"].includes(normalized)) return true;
  return DANGEROUS_PATH_PATTERNS.some((pat) => pat.test(normalized));
}

function looksLikeLiteralSecret(value) {
  const s = String(value);
  if (!s || s.startsWith("${")) return false;
  if (s.startsWith("ghp_") || s.startsWith("github_pat_")) return true;
  if (s.length >= 20 && /[a-zA-Z0-9]{20,}/.test(s)) {
    if (!s.includes("env:") && !s.includes("GITHUB_TOKEN")) return true;
  }
  return false;
}

function collectPathArgs(server) {
  const args = server.args;
  if (!Array.isArray(args)) return [];
  return args.filter((a) => typeof a === "string" || typeof a === "number");
}

function summarizeServer(name, cfg) {
  const hasCommand = typeof cfg.command === "string" && cfg.command.length > 0;
  const argCount = Array.isArray(cfg.args) ? cfg.args.length : 0;
  const envKeyCount =
    cfg.env && typeof cfg.env === "object"
      ? Object.keys(cfg.env).length
      : 0;
  return `  - ${name}: command_configured=${hasCommand} arg_count=${argCount} env_key_count=${envKeyCount}`;
}

function main() {
  const configIndex = process.argv.indexOf("--config");
  const configArg = configIndex >= 0 ? process.argv[configIndex + 1] : null;
  if (configIndex >= 0 && !configArg) {
    console.error("ERROR: --config requires a path");
    process.exit(1);
  }
  const mcpPath = configArg ? path.resolve(configArg) : DEFAULT_MCP_PATH;
  console.log("MCP config: selected");

  if (!fs.existsSync(mcpPath)) {
    console.error("ERROR: .cursor/mcp.json not found");
    process.exit(1);
  }

  let data;
  try {
    data = JSON.parse(fs.readFileSync(mcpPath, "utf8"));
  } catch (err) {
    console.error(`ERROR: invalid JSON: ${err.message}`);
    process.exit(1);
  }

  const servers = data.mcpServers;
  if (!servers || typeof servers !== "object" || Array.isArray(servers)) {
    console.error("ERROR: mcpServers must be an object");
    process.exit(1);
  }

  const names = Object.keys(servers).sort();
  console.log(`Servers (${names.length}): ${names.join(", ") || "(none)"}`);
  console.log("Summary:");
  for (const name of names) {
    const cfg = servers[name];
    if (cfg && typeof cfg === "object") {
      console.log(summarizeServer(name, cfg));
    }
  }

  const errors = [];
  const warnings = [];

  for (const [name, cfg] of Object.entries(servers)) {
    if (!cfg || typeof cfg !== "object") {
      errors.push(`${name}: config must be an object`);
      continue;
    }

    const isFilesystem =
      name === "filesystem" ||
      (Array.isArray(cfg.args) &&
        cfg.args.some((a) => String(a).includes("filesystem")));

    if (isFilesystem) {
      for (const arg of collectPathArgs(cfg)) {
        if (isDangerousPath(arg)) {
          errors.push(
            "filesystem: dangerous allowed path configured (use project workspace only)"
          );
        } else if (String(arg) === ".") {
          warnings.push(
            "filesystem: uses '.' — OK if Cursor cwd is repo root; else set single-repo absolute path locally"
          );
        }
      }
    }

    if (cfg.env && typeof cfg.env === "object") {
      for (const [key, val] of Object.entries(cfg.env)) {
        if (SENSITIVE_ENV_KEY.test(String(key)) && looksLikeLiteralSecret(val)) {
          errors.push(`${name}: env '${key}' looks like a hardcoded secret`);
        }
        if (
          String(val).startsWith("ghp_") ||
          String(val).startsWith("github_pat_")
        ) {
          errors.push(`${name}: GitHub token must not be committed`);
        }
      }
    }
  }

  if (names.includes("github")) {
    const gh = servers.github || {};
    const env = gh.env && typeof gh.env === "object" ? gh.env : {};
    const tokenRef = String(env.GITHUB_PERSONAL_ACCESS_TOKEN || "");
    if (!tokenRef.startsWith("${")) {
      warnings.push(
        "github: use env var reference (e.g. ${env:GITHUB_TOKEN}); set token in shell/Cursor env, never commit"
      );
    }
  }

  for (const w of warnings) console.log(`WARN: ${w}`);
  for (const e of errors) console.error(`ERROR: ${e}`);

  if (errors.length > 0) process.exit(1);

  console.log("OK: MCP config static check passed");
  console.log(
    `Supported candidates (not runtime claims): ${SUPPORTED_SERVERS.join(", ")}`
  );
  process.exit(0);
}

main();

File: scripts/check_mcp_config.py
#!/usr/bin/env python3
"""Lightweight static check for .cursor/mcp.json (no secrets printed)."""

from __future__ import annotations

import argparse
import json
import re
import sys
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1]
MCP_PATH = ROOT / ".cursor" / "mcp.json"

DANGEROUS_PATH_PATTERNS = [
    re.compile(r"^/$"),
    re.compile(r"^[A-Za-z]:\\?$", re.I),
    re.compile(r"^~/?$"),
    re.compile(r"^/Users/[^/]+/?$"),
    re.compile(r"^/home/[^/]+/?$"),
    re.compile(r"^C:\\Users\\[^\\]+/?$", re.I),
]

SENSITIVE_ENV_KEYS = re.compile(
    r"(token|secret|password|api[_-]?key|cookie|session)",
    re.I,
)

SUPPORTED_SERVERS = {
    "chrome-devtools",
    "context7",
    "filesystem",
    "github",
    "playwright",
}


def _collect_path_args(server: dict) -> list[str]:
    args = server.get("args") or []
    if not isinstance(args, list):
        return []
    return [str(a) for a in args if isinstance(a, (str, int, float))]


def _is_dangerous_path(path_arg: str) -> bool:
    normalized = path_arg.strip().replace("\\", "/")
    for pat in DANGEROUS_PATH_PATTERNS:
        if pat.match(normalized):
            return True
    if normalized in ("/", "//", "C:", "C:/"):
        return True
    return False


def _has_literal_secret(value: str) -> bool:
    if not value or value.startswith("${"):
        return False
    if len(value) >= 20 and re.search(r"[a-zA-Z0-9]{20,}", value):
        if "env:" not in value and "GITHUB_TOKEN" not in value:
            return True
    return False


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(description="Read-only MCP configuration check")
    parser.add_argument("--config", type=Path, default=MCP_PATH)
    return parser.parse_args()


def main() -> int:
    args = parse_args()
    mcp_path = args.config.resolve()
    print("MCP config: selected")
    if not mcp_path.exists():
        print("ERROR: .cursor/mcp.json not found")
        return 1

    try:
        data = json.loads(mcp_path.read_text(encoding="utf-8"))
    except json.JSONDecodeError as exc:
        print(f"ERROR: invalid JSON: {exc}")
        return 1

    servers = data.get("mcpServers") or {}
    if not isinstance(servers, dict):
        print("ERROR: mcpServers must be an object")
        return 1

    names = sorted(servers.keys())
    print(f"Servers ({len(names)}): {', '.join(names) or '(none)'}")

    errors: list[str] = []
    warnings: list[str] = []

    for name, cfg in servers.items():
        if not isinstance(cfg, dict):
            errors.append(f"{name}: config must be an object")
            continue

        if name == "filesystem" or "filesystem" in str(cfg.get("args", [])):
            for arg in _collect_path_args(cfg):
                if _is_dangerous_path(arg):
                    errors.append(
                        "filesystem: dangerous allowed path configured "
                        "(use project workspace only)"
                    )
                elif arg == ".":
                    warnings.append(
                        "filesystem: uses '.' — OK if Cursor cwd is repo root; "
                        "else set single-repo absolute path locally"
                    )

        env = cfg.get("env") or {}
        if isinstance(env, dict):
            for key, val in env.items():
                if SENSITIVE_ENV_KEYS.search(str(key)) and _has_literal_secret(
                    str(val)
                ):
                    errors.append(
                        f"{name}: env '{key}' looks like a hardcoded secret"
                    )
                if str(val).startswith("ghp_") or str(val).startswith("github_pat_"):
                    errors.append(f"{name}: GitHub token must not be committed")

    if "github" in names:
        gh = servers.get("github") or {}
        env = gh.get("env") if isinstance(gh, dict) else {}
        token_ref = ""
        if isinstance(env, dict):
            token_ref = str(env.get("GITHUB_PERSONAL_ACCESS_TOKEN", ""))
        if not token_ref.startswith("${env:"):
            warnings.append(
                "github: prefer ${env:GITHUB_TOKEN} in mcp.json; "
                "set GITHUB_TOKEN in shell/Cursor env (never commit token)"
            )

    for w in warnings:
        print(f"WARN: {w}")
    for e in errors:
        print(f"ERROR: {e}")

    if errors:
        return 1
    print("OK: MCP config static check passed")
    print(
        "Supported candidates (not runtime claims): "
        + ", ".join(sorted(SUPPORTED_SERVERS))
    )
    return 0


if __name__ == "__main__":
    sys.exit(main())
